src/main/java/de/tum/in/www1/artemis/service/RepositoryService.java

```diff
@@ -178,47 +179,76 @@ public Map<String, Boolean> getFilesWithInformationAboutChange(Repository reposi
         repository.setContent(null); // invalidate cache
         inputStream.close();
     }
 
-    private File checkIfFileExistsInRepository(Repository repository, String filename) throws FileAlreadyExistsException {
-        if (gitService.getFileByName(repository, filename).isPresent()) {
-            throw new FileAlreadyExistsException("file already exists");
-        }
-
-        File file = new File(Path.of(repository.getLocalPath().toString(), filename).toFile(), repository);
-        if (!repository.isValidFile(file)) {
-            throw new IllegalArgumentException();
-        }
-        return file;
-    }
-
     /**
      * Create a folder in a repository.
      *
      * @param repository  in which the folder should be created.
-     * @param folderName  of the folder to be created.
+     * @param folderPath  of the folder to be created.
      * @param inputStream byte representation of the folder to be created.
      * @throws IOException if the inputStream is corrupt, the folder can't be stored, the repository is unavailable, etc.
      */
-    public void createFolder(Repository repository, String folderName, InputStream inputStream) throws IOException {
-        checkIfFileExistsInRepository(repository, folderName);
-        Files.createDirectory(repository.getLocalPath().resolve(folderName));
+    public void createFolder(Repository repository, String folderPath, InputStream inputStream) throws IOException {
+        Path safePath = checkIfPathIsValidAndExistsAndReturnSafePath(repository, folderPath, false);
+        checkIfPathAndFileAreValidAndReturnSafeFile(repository, safePath);
+        Files.createDirectory(repository.getLocalPath().resolve(safePath));
         // We need to add an empty keep file so that the folder can be added to the git repository
-        File keep = new File(repository.getLocalPath().resolve(folderName).resolve(".keep"), repository);
+        File keep = new File(repository.getLocalPath().resolve(safePath).resolve(".keep"), repository);
         Files.copy(inputStream, keep.toPath(), StandardCopyOption.REPLACE_EXISTING);
         repository.setContent(null); // invalidate cache
         inputStream.close();
     }
 
+    private Path checkIfPathIsValidAndReturnSafePath(Repository repository, String path) {
+        String unescapedPath = StringEscapeUtils.unescapeJava(path);
+        Path unknownInputPath = Paths.get(unescapedPath).normalize();
+        Path absoluteRepositoryPath = repository.getLocalPath().normalize().toAbsolutePath();
+        Path absoluteInputPath = absoluteRepositoryPath.resolve(unknownInputPath).normalize();
+        if (!absoluteInputPath.startsWith(absoluteRepositoryPath)) {
+            throw new IllegalArgumentException("Path is not valid");
+        }
+        return unknownInputPath;
+    }
+
+    private Path checkIfPathIsValidAndExistsAndReturnSafePath(Repository repository, String path, boolean exists) {
```

Comment:
I'm not sure the method name is correct here. I don't think the method checks if the path exists, rather it is specified in the parameter if it supposed to exist. Or am I mistaken?